compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5939,15 +5952,85 @@ TR::Register *OMR::Power::TreeEvaluator::setmemoryEvaluator(TR::Node *node, TR::
    TR::LabelSymbol * label1aligned =  generateLabelSymbol(cg);
 
    TR::RegisterDependencyConditions *conditions;
-   int32_t numDeps = 5;
+   int32_t numDeps = 7;
    conditions = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(numDeps, numDeps, cg->trMemory());
    TR::Register *cndReg = cg->allocateRegister(TR_CCR);
    TR::addDependency(conditions, cndReg, TR::RealRegister::cr0, TR_CCR, cg);
-   TR::addDependency(conditions, dstAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(conditions, dstBaseAddrReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(conditions, dstOffsetReg, TR::RealRegister::NoReg, TR_GPR, cg);
    TR::addDependency(conditions, lengthReg, TR::RealRegister::NoReg, TR_GPR, cg);
    TR::addDependency(conditions, valueReg, TR::RealRegister::NoReg, TR_GPR, cg);
-   TR::Register * tempReg = cg->allocateRegister();
-   TR::addDependency(conditions, tempReg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::Register * temp1Reg = cg->allocateRegister();
+   TR::Register * temp2Reg = cg->allocateRegister();
+   TR::addDependency(conditions, temp1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(conditions, temp2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+
+#if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+   // When using balanced GC policy with offheap allocation enabled, there are three possible cases:
+   // 1.) The object at dstBaseAddr is known to be a non-array object at compile time. In this scenario, no arrayCHK is
+   //     generated, and no adjustments are made to dstBaseAddr or dstOffset. The behavior in this case should be identical
+   //     to that under gencon GC policy.
+   // 2.) The object at dstBaseAddr is known to be an array at compile time. In this scenario, no arrayCHK is generated, but
+   //     the dstBaseAddr and dstOffset with be adjusted as needed for offheap.
+   // 3.) The type of the object at dstBaseAddr is unknown at compile time. In this scenario, a runtime arrayCHK will generated,
+   //     with two possible outcomes: if the object is an array, the dstBaseAddr and dstOffset will be adjusted, and if not,
+   //     no adjustments will be made.
+
+   //check dstBaseAddrNode type at compile time
+   int length;
+   const char *objTypeSig = dstBaseAddrNode->getSymbolReference()->getTypeSignature(length);
+
+   //generate arrayCHK in case (3) only
+   bool arrayCheckNeeded = TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit() &&
+                           (objTypeSig == NULL || strstr(objTypeSig, "Ljava/lang/Object"));
```

Comment:
again, strstr only compares the starting part (not the whole string) as mentioned in the other PR.  it can return false-positive if the objType is something like  "`Ljava/lang/ObjectFoo`" or "`Ljava/lang/ObjectMyType`"